extensions/notification/utils/config.js

```diff
@@ -43,42 +43,45 @@ function validateEntryWithFilter(filterRules, entry) {
 }
 
 /**
- * Validate event against bucket notification configuration event type
- * @param  {Object[]} bnConfigs - Array of bucket specific notification
+ * Validate event against bucket queue configuration event type
+ * @param  {Object[]} bnConfigs - Array of bucket specific queue
  * configurations
  * @param  {string} event - Type of event
- * @return {undefined|Object} A bucket notification configuration if found
+ * @return {undefined|Object[]} Bucket queue configurations if found
  */
 function validateConfigsEvent(bnConfigs, event) {
-    const config = bnConfigs.filter(config => config.events.includes(event));
-    return config.length > 0 ? config[0] : undefined;
+    const configs = bnConfigs.filter(config => config.events.includes(event));
+    return configs.length > 0 ? configs : undefined;
 }
 
 /**
  * Validates an entry from log against bucket notification configurations to see
  * if the entry has to be published. Validations include, bucket specific
  * configuration check, event type check, object name specific filter checks
- * @param  {Object[]} bnConfigs - Array of bucket notifications configuration
+ * @param  {Object} bnConfig - Bucket notification configuration
  * @param  {Object} entry - An entry from the log
  * @return {boolean} true if event qualifies for notification
  */
-function validateEntry(bnConfigs, entry) {
+function validateEntry(bnConfig, entry) {
     const { bucket, type } = entry;
-    // check if the entry belongs to any bucket in the configuration
-    if (bnConfigs.has(bucket)) {
-        const bnConfig = bnConfigs.get(bucket);
-        // check if the event type matches
-        const qConfig = validateConfigsEvent(bnConfig, type);
-        if (qConfig) {
-            // check if there are filters and the object matches it
-            const configFilter = qConfig.filter;
-            if (configFilter && configFilter.filterRules) {
-                return validateEntryWithFilter(configFilter.filterRules, entry);
-            }
-            return true;
-        }
+    const notifConf = bnConfig.notificationConfiguration;
+    // check if the entry belongs to the bucket in the configuration
+    if (bucket !== bnConfig.bucket) {
         return false;
     }
+    // check if the event type matches
+    const qConfigs = validateConfigsEvent(notifConf.queueConfig, type);
+    if (qConfigs) {
+        // check if there are filters for any config and the object's key
+        // satisfies any filter
+        const qFilters
+            = qConfigs.filter(c => c.filterRules && c.filterRules.length > 0);
+        if (qFilters.length > 0) {
+            return qConfigs.some(
+                c => validateEntryWithFilter(c.filterRules, entry));
```

Comment:
I don't think the logic is right as `c.filterRules` could be undefined here (does not seem handled by `validateEntryWithFilter`).

```suggestion
                return qConfigs.some(c.filterRules && c => validateEntryWithFilter(c.filterRules, entry));
```
And remove the previous `qFilter = qConfig...` + `if (qFilters.length > 0)` checks.

Reply:
The check here is to see if a bucket notification configuration has filters (filters are optional), if they don't (which is checked by `if (qFilters.length > 0)` ) `validateEntryWithFilter` will be skipped.

Reply:
What if you have `qConfig` contain two entries like:

```
[{ filterRules: [{ some rule }] }, { /* no filterRules here */ }]
```

I may be wrong but it seems to be this would lead to `qFilters.length > 0`, then `validateEntryWithFilter` would raise an exception on the second entry because there is no `filterRules` in the second entry.

Reply:
`qFilters` will contain only configs with filters in them, ref line #77, so `validateEntryWithFilter` would not be validating a config without filters.

Reply:
But you're doing `qConfigs.some` not `qFilters.some`, maybe that is a typo then?

Reply:
In any case, make sure this situation is covered in unit tests.

Reply:
New unit tests added to cover the condition discussed.